File: 2021/haskell/app/D04.hs
module D04 where

import           Data.List
import           Data.List.Split
import           Utils


main :: IO ()
main = defaultMain defaultFile parse part1 part2

defaultFile :: String
defaultFile = "../input/d04.txt"

-- A bingo grid, vectorised.
type Grid = [Int]

at :: Grid -> Int -> Int -> Int
at g x y = g !! (y * 5 + x)

wins :: [Int] -> Grid -> Bool
wins hist g = or [ hit [ at g i j | j <- [0 .. 4] ] | i <- [0 .. 4] ]
    || or [ hit [ at g i j | i <- [0 .. 4] ] | j <- [0 .. 4] ]
    -- || hit [at g i i | i <- [0..4]] -- Diagonals don't count?
    -- || hit [at g i (4-i) | i <- [0..4]]
    where hit = all (`elem` hist)

score :: [Int] -> Grid -> Int
score hist g = sum [ x | x <- g, x `notElem` hist ] * last hist

parse :: String -> ([Int], [Grid])
parse s =
    let (xs : _ : rest) = lines s
        grids = map (map read . concatMap words) $ splitOn [""] rest
    in  (map read $ splitOn "," xs, grids)

-- Iterates over the numbers, returns a list of pairs: (iter i, grids that completed on iter i).
iter :: [Int] -> [Grid] -> [(Int, [Grid])]
iter ns gs = iter' ns gs 1
  where
    iter' ns' gs' i | i == length ns' = []
                    | otherwise       = (i, won) : iter' ns' ongoing (i + 1)
        where (won, ongoing) = partition (wins (take i ns')) gs'

part1 :: ([Int], [Grid]) -> Int
part1 (ns, gs) = score (take i ns) (head firstWinGroup)
    where (i, firstWinGroup) = firstBy (not . null . snd) $ iter ns gs

part2 :: ([Int], [Grid]) -> Int
part2 (ns, gs) = score (take i ns) (head firstWinGroup)
    where (i, firstWinGroup) = lastBy (not . null . snd) $ iter ns gs

File: 2021/rust/src/bin/d04.rs
use std::fs;

type Grid = Vec<u32>;

fn main() {
    let filename = "../input/d04.txt";

    let contents = fs::read_to_string(filename).unwrap();
    let (nums, grids) = parse(contents);

    println!("part1: {}", part1(&nums, &grids));
    println!("part2: {}", part2(&nums, &grids));
}

fn parse(input: String) -> (Vec<u32>, Vec<Grid>) {
    let mut sections = input.split("\n\n");
    let nums: Vec<u32> = sections
        .next()
        .unwrap()
        .split(',')
        .map(|x| x.parse::<u32>().unwrap())
        .collect();
    let grids = sections
        .map(|g| {
            g.split(char::is_whitespace)
                .filter(|x| !x.is_empty())
                .map(|x| x.parse::<u32>().unwrap())
                .collect::<Vec<_>>()
        })
        .collect::<Vec<_>>();
    (nums, grids)
}

fn part1(nums: &Vec<u32>, grids: &Vec<Grid>) -> u32 {
    let ngrids = grids.len();
    let mut solved = vec![false; ngrids];
    let mut marks = vec![vec![false; 5 * 5]; ngrids];
    for x in nums {
        for (gi, g) in grids.iter().enumerate() {
            if !solved[gi] && g.contains(&x) {
                let gii = g.iter().position(|y| y == x).unwrap();
                marks[gi][gii] = true;
                if check_win(&marks[gi], gii) {
                    solved[gi] = true;
                    return grid_score(g, &marks[gi], x);
                }
            }
        }
    }
    unreachable!();
}

fn part2(nums: &Vec<u32>, grids: &Vec<Grid>) -> u32 {
    let ngrids = grids.len();
    let mut solved_grids = 0usize;
    let mut solved = vec![false; ngrids];
    let mut marks = vec![vec![false; 5 * 5]; ngrids];
    for x in nums {
        for (gi, g) in grids.iter().enumerate() {
            if !solved[gi] && g.contains(&x) {
                let gii = g.iter().position(|y| y == x).unwrap();
                marks[gi][gii] = true;
                if check_win(&marks[gi], gii) {
                    solved[gi] = true;
                    solved_grids += 1;
                    if solved_grids == ngrids {
                        return grid_score(g, &marks[gi], x);
                    }
                }
            }
        }
    }
    unreachable!();
}

fn at(i: usize, j: usize) -> usize {
    return j * 5 + i;
}
fn index2(i: usize) -> (usize, usize) {
    return (i % 5, i / 5);
}

fn check_win(mark: &Vec<bool>, last_update: usize) -> bool {
    let (x, y) = index2(last_update);

    return (0..=4).all(|i| mark[at(x, i)]) || (0..=4).all(|i| mark[at(i, y)]);
}

fn grid_score(grid: &Grid, mark: &Vec<bool>, last: &u32) -> u32 {
    return grid
        .iter()
        .zip(mark)
        .map(|(x, m)| x * (!m as u32))
        .sum::<u32>()
        * last;
}
